ballista/core/src/physical_optimizer/task_group.rs

```diff
@@ -81,9 +82,10 @@ impl OptimizeTaskGroup {
 
             Ok(Transformed::Yes(new_plan))
         } else if node.as_any().is::<UnionExec>()
-            && children
-                .iter()
-                .all(|child| child.as_any().is::<CoalesceTasksExec>())
+            || is_hash_join_no_partitioning(node.as_ref())
+                && children
+                    .iter()
+                    .all(|child| child.as_any().is::<CoalesceTasksExec>())
```

Comment:
Looks like you might need parens around `(isUnion || isHashJoin) && allChildrenAreCoalesce`

Reply:
Ah yeah 😜